nat-lab/tests/telio.py

```diff
@@ -213,57 +213,76 @@ async def notify_peer_state(
         public_key: str,
         states: List[State],
         paths: List[PathType],
-        new_event_explicitly: bool = False,
     ) -> None:
-        def _get_events() -> Set[PeerInfo]:
-            return {
-                event
-                for event in self._peer_state_events
-                if event.public_key == public_key
-                and event.state in states
-                and event.path in paths
-            }
+        while True:
+            peer = self.get_peer_info(public_key)
+            if peer and peer.path in paths and peer.state in states:
+                return
+            await asyncio.sleep(0.01)
+
+    async def notify_peer_event(
+        self,
+        public_key: str,
+        states: List[State],
+        path: PathType,
+        event: Optional[asyncio.Event] = None,
+    ) -> None:
+        def _get_events() -> List[PeerInfo]:
+            return [
+                peer
+                for peer in self._peer_state_events
+                if peer
+                and peer.public_key == public_key
+                and peer.path == path
+                and peer.state in states
+            ]
 
         old_events = _get_events()
+
+        if event:
+            event.set()
```

Comment:
I think you missed the point. Event is there to synchronize tasks, after we initialize old_events the event is set() to inform the main thread that the event waiting task is ready (we got the historic events, no dataraces will happen), giving green light to execute the test actions. Therefore, moving it to While loop is trivial and adds unnecessary cpu iterations. 

You can do `future = run_async(notifiy_peer_event(...))`,  however how do you ensure that notifiy_peer_event() have started and old_events been fetched before starting test actions? You could await on that future but that would block. So I came up with Event() because it truly launchs the task asynchrounisly and confirms we are ready to move on.

Reply:
Just to add that, we do not want to wait for the function return, but for the function signal that the old events have been fetched, allowing us to start the test actions without eventual dataraces.

Reply:
Can you show some use cases of it, cuz I cannot wrap my head around, why is it actually needed and why it is not used anywhere. :confused:

Reply:
Oops, it seems I did some last minute changes that I didn't remember, sorry about that, you are right.

Reply:
So currently we are not using them, I'll have them removed. 

The thought behind it was to avoid dataraces. For instance, if we have an event that we would like to catch but it was logged so quickly after the test action that we couldn't call wait_for_event() in time, missing it.

Reply:
It technically still can happen, see: https://github.com/NordSecurity/libtelio/pull/41/files#diff-778fbedd2d660ca2c2322076ce67ff83072b808e239308da0bbfc3f130add155R331